File: code/ch04/4.4.1.GameOfLife.cpp
class Solution {
private:
    int get_neighbor_count(int i, int j, vector<vector<int>>& board) {
        int top = std::max(0, i - 1);
        int bottom = std::min<int>(board.size() - 1, i + 1);
        int left = std::max(0, j - 1);
        int right = std::min<int>(board[0].size() - 1, j + 1);
        
        int count = 0;
        for (int x = top; x < bottom + 1; x++)
            for (int y = left; y < right + 1; y++)
                if (board[x][y] == 1 || board[x][y] == -1)
                    count += 1;
        
        return count;
    }
public:
    void gameOfLife(vector<vector<int>>& board) {
        for (int i = 0; i < board.size(); i++) {
            for (int j = 0; j < board[0].size(); j++) {
                int res = get_neighbor_count(i, j, board);
                if (board[i][j] == 1 && (res == 3 || res == 4))
                    board[i][j] = 1;
                else if (board[i][j] == 1)
                    board[i][j] = -1;
                else if (board[i][j] == 0 && res == 3)
                    board[i][j] = -2;
            }
        }
        
        for (int i = 0; i < board.size(); i++) {
            for (int j = 0; j < board[0].size(); j++) {
                if (board[i][j] == -2)
                    board[i][j] = 1;
                else if (board[i][j] == -1)
                    board[i][j] = 0;
            }
        }
    }
};

File: code/ch04/4.4.1.GameOfLife.py
from typing import List


class Solution:
    def gameOfLife(self, board: List[List[int]]) -> None:
        def get_neighbor_count(i, j, board):
            top = max(0, i - 1)
            bottom = min(len(board) - 1, i + 1)
            left = max(0, j - 1)
            right = min(len(board[0]) - 1, j + 1)

            count = 0
            for x in range(top, bottom + 1):
                for y in range(left, right + 1):
                    if board[x][y] == 1 or board[x][y] == -1:
                        count += 1
            return count

        for i in range(len(board)):
            for j in range(len(board[0])):
                res = get_neighbor_count(i, j, board)
                if board[i][j] == 1 and res in [3, 4]:
                    board[i][j] = 1
                elif board[i][j] == 1:
                    board[i][j] = -1
                elif board[i][j] == 0 and res == 3:
                    board[i][j] = -2

        for i in range(len(board)):
            for j in range(len(board[0])):
                if board[i][j] == -2:
                    board[i][j] = 1
                elif board[i][j] == -1:
                    board[i][j] = 0
